Scripts/Runtime/Entities/Lifecycle/EntitySpawnSystem.cs

```diff
@@ -30,22 +31,47 @@ namespace Anvil.Unity.DOTS.Entities
     [UseCommandBufferSystem(typeof(EndSimulationEntityCommandBufferSystem))]
     public partial class EntitySpawnSystem : AbstractAnvilSystemBase
     {
+        private readonly AccessControlledValue<NativeParallelHashMap<long, Entity>> m_EntityPrototypes;
+        private readonly HashSet<EntitySpawner> m_ActiveSpawners;
+        private readonly List<EntitySpawner> m_PendingReleaseSpawners;
+        private readonly uint m_SystemID;
+
+        private readonly Queue<int> m_InstanceIDQueue;
+        private int m_LargestInstanceIDIssued;
+
         private EntityCommandBufferSystem m_CommandBufferSystem;
         private NativeParallelHashMap<long, EntityArchetype> m_EntityArchetypes;
 
-        private readonly Dictionary<Type, IEntitySpawner> m_EntitySpawners;
-        private readonly HashSet<IEntitySpawner> m_ActiveEntitySpawners;
-        private readonly AccessControlledValue<NativeParallelHashMap<long, Entity>> m_EntityPrototypes;
-
         public EntitySpawnSystem()
         {
-            m_EntitySpawners = new Dictionary<Type, IEntitySpawner>();
-            m_ActiveEntitySpawners = new HashSet<IEntitySpawner>();
+            m_ActiveSpawners = new HashSet<EntitySpawner>();
+            m_PendingReleaseSpawners = new List<EntitySpawner>();
+            m_InstanceIDQueue = new Queue<int>();
+            m_LargestInstanceIDIssued = -1;
+
             m_EntityArchetypes = new NativeParallelHashMap<long, EntityArchetype>(ChunkUtil.MaxElementsPerChunk<EntityArchetype>(), Allocator.Persistent);
             m_EntityPrototypes = new AccessControlledValue<NativeParallelHashMap<long, Entity>>(
                 new NativeParallelHashMap<long, Entity>(
                     ChunkUtil.MaxElementsPerChunk<Entity>(),
                     Allocator.Persistent));
+            m_SystemID = s_EntitySpawnSystemIDProvider.GetNextID();
+        }
+
+        private int GetNextInstanceID()
+        {
+            //We'd use a Pool<T> here but it doesn't work with primitives so we get around that with the Queue/largest
+            //instance id issued combo.
+            //This way we don't run out of IDs when acquiring every frame but we also don't need to force the call site
+            //to manage and store an ID.
+            
+            //If we have an ID free to use in the queue, just get it
+            if (m_InstanceIDQueue.Count > 0)
+            {
+                return m_InstanceIDQueue.Dequeue();
+            }
```

Comment:
Should we have a `PooledIDProvider` or use `Pool<T>` for this? 

I looked at the `Pool<T>` implementation and it would be trivial to support value types.

Reply:
Yes I originally wanted to use a Pool but couldn't and it was a tangent to go down. 

Added a TODO and linked to: https://github.com/decline-cookies/anvil-csharp-core/issues/147

Reply:
Looks like all you have to do is change one line to support value types in `Pool<T>` that defaults the acquired value to `null`